Block Type: Add selectors to block.json schema

aaronrobertshaw opened the pull request:
This PR adds support for the new `selectors` property. It adds it to the allowed metadata when registering a block type, makes the `WP_Block_Type` class aware of it, exposes it through the block types REST API, and the `get_block_editor_server_block_settings` function.

There is a corresponding Gutenberg PR (https://github.com/WordPress/gutenberg/pull/46496) that stabilizes the block.json `selectors` API and in part depends upon the changes here.

Trac ticket: https://core.trac.wordpress.org/ticket/57585

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

aaronrobertshaw committed:
Block Type: Add selectors to block.json schema

gziolo commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -117,6 +117,14 @@ class WP_Block_Type {
+	 * @since 6.2.0
+	 * @var array|null
+	 */
+	public $selectors = null;
```
As noted in https://github.com/WordPress/gutenberg/pull/46496#pullrequestreview-1295401223, we need to decide whether the default value should be `null` or and empty array.

aaronrobertshaw replied:
I've updated the default to be an empty array so at least for now, the Gutenberg PR, its docs, and this patch are consistent.

gziolo commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -117,6 +117,14 @@ class WP_Block_Type {
+	/**
+	 * Custom CSS selectors for theme.json style generation.
+	 *
+	 * @since 6.2.0
```
It's probably too late so:
```suggestion
	 * @since 6.3.0
```

and in other places.

aaronrobertshaw replied:
Thanks, agreed. It definitely missed 6.2 🙂

gziolo commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -268,6 +277,7 @@ class WP_Block_Type {
 	 *     @type string|null   $textdomain               The translation textdomain.
 	 *     @type array[]       $styles                   Alternative block styles.
 	 *     @type array[]       $variations               Block variations.
+	 *     @type array|null    $selectors                Custom CSS selectors for theme.json style generation.
```
Does `null` have any special meaning here? I'm still not sure why we set `null` in some places but map it to an empty object/array on the client. It might be related to some backward compatibility considerations in places where `null` doesn't have a special role like for `example` where it disables block preview in the inserter.

aaronrobertshaw replied:
I don't think there is a special meaning for `null` here. It was more to keep it inline with the `supports` property. 

For now, I've tweaked the checks for the `selectors` property in https://github.com/WordPress/gutenberg/pull/46496 and updated this PR so that the default is an empty array. Happy to change this again though if we decide otherwise.

gziolo commented:
It's nearly ready from my perspective. I left my comments, and deciding on the default value is the only important thing.

aaronrobertshaw committed:
Update selectors default and since

gziolo reviewed (approved):
This is ready to go. We will need to wait with the commit until changes in the Gutenberg plugin land and the release cycle for WordPress 6.2 is finished. Feel free to remind me to to commit it to WordPress core when the time comes 😄

aaronrobertshaw committed:
Add editor only selectors to block type

aaronrobertshaw commented:
@gziolo there have been a few changes here. I've added a second property to block types for editor-only selectors. Some background for this change can be found over on the related Gutenberg PR (https://github.com/WordPress/gutenberg/pull/46496#issuecomment-1464031044). 

Essentially, we went with the second property to match the existing `style` / `editorStyle` block type properties, so we'll offer `selectors` / `editorSelectors`.

aaronrobertshaw committed:
Fix editor selectors position in expected block settings

gziolo commented:
> Essentially, we went with the second property to match the existing `style` / `editorStyle` block type properties, so we'll offer `selectors` / `editorSelectors`.

Sounds good, and the code changes look correct. I'm not sure why e2e tests failed on CI but I restarted the job to see if that was only an incident.

aaronrobertshaw commented:
>  I'm not sure why e2e tests failed on CI but I restarted the job to see if that was only an incident.

Seems like you had the magic touch 🪄

I'd already re-run it twice before rebasing the PR and having it fail again.

hellofromtonya commented on tests/phpunit/tests/blocks/register.php:
```diff
@@ -515,6 +515,14 @@ public function test_block_registers_with_metadata_fixture() {
 			$result->provides_context
 		);
 		$this->assertSameSets( array( 'groupId' ), $result->uses_context );
+		$this->assertSame(
+			array( 'root' => '.editor-only.wp-block-notice' ),
+			$result->editor_selectors
+		);
+		$this->assertSame(
+			array( 'root' => '.wp-block-notice' ),
+			$result->selectors
+		);
```
Please add assertion messages to each of these `assertSame()`. Why? If any fail, the message will identify which of the assertions failed.

hellofromtonya replied:
Also add `@ticket 57585` to the test method.

hellofromtonya commented on tests/phpunit/tests/rest-api/rest-block-type-controller.php:
```diff
@@ -216,6 +216,8 @@ public function test_get_item_invalid() {
 			'example'          => 'invalid_example',
 			'parent'           => 'invalid_parent',
 			'ancestor'         => 'invalid_ancestor',
+			'editor_selectors' => 'invalid_editor_selectors',
+			'selectors'        => 'invalid_selectors',
```
Add `@ticket 57585` to the test method to identify this change for tacking.

hellofromtonya commented on tests/phpunit/tests/rest-api/rest-block-type-controller.php:
```diff
@@ -541,7 +549,7 @@ public function test_get_item_schema() {
 		$data       = $response->get_data();
 		$properties = $data['schema']['properties'];
-		$this->assertCount( 28, $properties );
+		$this->assertCount( 30, $properties );
```
Add `@ticket 57585` to the test method to identify this change for tacking.

hellofromtonya replied:
And add a assertion failure message too please.

hellofromtonya commented on tests/phpunit/tests/rest-api/rest-block-type-controller.php:
```diff
@@ -551,6 +559,8 @@ public function test_get_item_schema() {
 		$this->assertArrayHasKey( 'textdomain', $properties );
 		$this->assertArrayHasKey( 'name', $properties );
 		$this->assertArrayHasKey( 'attributes', $properties );
+		$this->assertArrayHasKey( 'editor_selectors', $properties );
+		$this->assertArrayHasKey( 'selectors', $properties );
```
* Add @ticket 57585 to the test method to identify this change for tacking.
* Add an assertion failure message to each assertion please.

hellofromtonya commented on src/wp-admin/includes/post.php:
```diff
@@ -2192,6 +2192,8 @@ function get_block_editor_server_block_settings() {
 		'provides_context' => 'providesContext',
 		'uses_context'     => 'usesContext',
+		'selectors'        => 'selectors',
+		'editor_selectors' => 'editorSelectors',
```
In the function's DocBlock, please add:

```php
 * @since 6.3.0 Added `'selectors'` and `'editor_selectors'` fields.
```

hellofromtonya commented on src/wp-includes/blocks.php:
```diff
@@ -382,6 +382,8 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 		'attributes'      => 'attributes',
 		'providesContext' => 'provides_context',
 		'usesContext'     => 'uses_context',
+		'selectors'       => 'selectors',
+		'editorSelectors' => 'editor_selectors',
```
In the function's DocBlock, please add:

```php
 * @since 6.3.0 Added `'selectors'` and `'editor_selectors'` fields.
```

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php:
```diff
@@ -278,6 +278,8 @@ public function prepare_item_for_response( $item, $request ) {
 				'ancestor',
 				'provides_context',
 				'uses_context',
+				'selectors',
+				'editor_selectors',
```
In the function's DocBlock, please add:

```php
 * @since 6.3.0 Added `'selectors'` and `'editor_selectors'` fields.
```

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php:
```diff
@@ -518,6 +520,22 @@ public function get_item_schema() {
+					'context'     => array( 'embed', 'view', 'edit' ),
+					'readonly'    => true,
+				),
+				'editor_selectors'      => array(
```
In the function's DocBlock, please add:

```php
 * @since 6.3.0 Added `'selectors'` and `'editor_selectors'` fields.
```

aaronrobertshaw committed:
Update DocBlocks, add ticket refs and assertion messages

aaronrobertshaw commented:
👍 Thanks for the feedback @hellofromtonya. I believe I've addressed all of that now in https://github.com/WordPress/wordpress-develop/pull/3942/commits/82d1762b4bc5a9e8eaf114a58022f74302f7a973

aaronrobertshaw committed:
Revert "Fix editor selectors position in expected block settings"

This reverts commit 635ffee4c33e931ed7243208b88ef1e7f62040ea.

aaronrobertshaw committed:
Revert "Add editor only selectors to block type"

aaronrobertshaw committed:
Remove references to editor_selectors in DocBlocks

aaronrobertshaw commented:
There were [some concerns](https://github.com/WordPress/gutenberg/pull/46496#discussion_r1149645412) raised around introducing support for editor-only selectors on the Gutenberg PR. For the time being, we'll remove that editor-only support so it can be discussed further and added at a future date. 

As such, I've reverted the addition of editor selectors to block types here.

gziolo commented:
@aaronrobertshaw, it's been a while since all the changes landed in the Gutenberg plugin. Do you think it's good to go for WordPress core? The code changes look good.

aaronrobertshaw commented:
Thanks for the nudge @gziolo!

After the removal of the editor-only selectors from this API, it doesn't seem like there is a great appetite yet to see them return. I suspect it might take some time to find solutions to the issues facing editor-only selectors. 

Given that, I believe the changes here are stable and as you note the Selectors API landed in Gutenberg 15.5 so they've in the wild for a little while.

There's no objections from my end to see this land in core 🙂

gziolo commented:
Committed with https://core.trac.wordpress.org/changeset/55673.
